pkg/services/datasources/service/secrets_mig.go

```diff
@@ -0,0 +1,52 @@
+package service
+
+import (
+	"context"
+
+	"github.com/grafana/grafana/pkg/services/datasources"
+)
+
+type DataSourceSecretMigrationService struct {
+	dataSourcesService datasources.DataSourceService
+}
+
+func ProvideDataSourceMigrationService(dataSourcesService datasources.DataSourceService) *DataSourceSecretMigrationService {
+	return &DataSourceSecretMigrationService{
+		dataSourcesService: dataSourcesService,
+	}
+}
+
+func (s *DataSourceSecretMigrationService) Migrate(ctx context.Context) error {
+	query := &datasources.GetAllDataSourcesQuery{}
+	err := s.dataSourcesService.GetAllDataSources(ctx, query)
+	if err != nil {
+		return err
+	}
+
+	for _, ds := range query.Result {
+		hasMigration := ds.JsonData.Get("secretMigrationComplete").MustBool()
+		if !hasMigration {
+			secureJsonData, err := s.dataSourcesService.DecryptedValues(ctx, ds)
+			if err != nil {
+				return err
+			}
+
+			ds.JsonData.Set("secretMigrationComplete", true)
```

Comment:
That is possible now, since the `decryptLegacySecrets` function is not exposed anymore we use `DecryptedValues` to retrive secrets, that prevents us from trying to migrate deleted secrets again and deleting already migrated secrets, in this case we don't need to store the migration state for each data source and we can just use the kvstore to store the migration state for the whole migration.

We could just set the secret migration complete on the kvstore, I don't think it's necessary to have 2 keys for the compatiblity removal since the flag can be used directly.

Reply:
I believe the second key wasn't necessary. I used a `secretMigrationStatus` key, on the `datasource` namespace and orgId 0. The values can be `""` if it's not set (migration was never run), `"compatible"` if migration was run with compatibility enabled, or `"complete"` if migration was run with compatibility disabled.

Migration is run based on the following confitions:
```go
	disableSecretsCompatibility := s.features.IsEnabled("disableSecretsCompatibility")
	needCompatibility := migrationStatus != "compatible" && !disableSecretsCompatibility
	needMigration := migrationStatus != "complete" && disableSecretsCompatibility

	if needCompatibility || needMigration {
	  // run migration
	}
```

As done on https://github.com/grafana/grafana/pull/50463/commits/254c5585d430ee5e021f1f830d430ae600844c05

Reply:
Sounds like a nice approach :+1: